### PR TITLE
Enable query and node heartbeats to resource manager

### PR DESCRIPTION
**Pull request by tdcmeehan**
Test plan - Unit tests, shadow tests, verifier

Extracted from: #15479

Depends on: #15892

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/InMemoryNodeManager.java [hunk 1]
```diff
@@ -132,7 +131,7 @@ public InternalNode getCurrentNode()
     public Set<InternalNode> getResourceManagers()
     {
         // always use localNode as resource manager
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/metadata/InMemoryNodeManager.java:133]: - The comment is out of date
- Do we need to completely remove localNode given we have dedicated RM tag now?

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RandomResourceManagerAddressSelector.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    @Override
+    public Optional<SimpleAddress> selectAddress(Optional<String> addressSelectionContext)
+    {
+        if (addressSelectionContext.isPresent()) {
+            return addressSelectionContext
+                    .map(HostAndPort::fromString)
+                    .map(SimpleAddress::new);
+        }
+        List<HostAndPort> resourceManagers = internalNodeManager.getResourceManagers()
+                .stream()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RandomResourceManagerAddressSelector.java:63]: merge `.stream()` to the previous line

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RandomResourceManagerAddressSelector.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                .filter(node -> node.getThriftPort().isPresent())
+                .map(resourceManagerNode -> {
+                    HostAddress hostAndPort = resourceManagerNode.getHostAndPort();
+                    return HostAndPort.fromParts(hostAndPort.getHostText(), resourceManagerNode.getThriftPort().getAsInt());
+                })
+                .collect(toImmutableList());
+        return hostSelector.apply(resourceManagers)
+                .map(SimpleAddress::new);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RandomResourceManagerAddressSelector.java:71]: merge into one line

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        QueryId queryId = queryExecution.getBasicQueryInfo().getQueryId();
+        queries.computeIfAbsent(queryId, unused -> new PeriodicTaskExecutor(
+                resourceManagerConfig.getQueryHeartbeatInterval().toMillis(),
+                executor,
+                () -> sendQueryHeartbeat(queryExecution)));
+        queryExecution.addStateChangeListener(newState -> {
+            if (newState.isDone()) {
+                queries.computeIfPresent(queryId, (unused, queryHeartbeatSender) -> {
+                    queryHeartbeatSender.forceRun();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java:109]: This is a blocking query on query completion. Will this cause query completion time to be too long (~10ms) causing long finishing time?

File: presto-main/src/main/java/com/facebook/presto/util/PeriodicTaskExecutor.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    private final long delayTargetMillis;
+    private final ScheduledExecutorService executor;
+    private final Runnable runnable;
+    private final LongUnaryOperator nextDelayFunction;
+
+    @GuardedBy("this")
+    private long delayMillis;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+    @GuardedBy("this")
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/util/PeriodicTaskExecutor.java:40]: Possible to make these calls non-synchronized I guess?

File: presto-main/src/test/java/com/facebook/presto/resourcemanager/MockResourceManagerClient.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import com.facebook.presto.spi.memory.MemoryPoolId;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class MockResourceManagerClient
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/resourcemanager/MockResourceManagerClient.java:30]: `TestingResourceManagerClient `

File: presto-main/src/test/java/com/facebook/presto/util/TestPeriodicTaskExecutor.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        CountDownLatch latch = new CountDownLatch(3);
+        Runnable runnable = latch::countDown;
+
+        try (PeriodicTaskExecutor executor = new PeriodicTaskExecutor(SECONDS.toMillis(durationBetweenTicksInSeconds), executorService, runnable, i -> i)) {
+            Stopwatch stopwatch = Stopwatch.createStarted();
+            latch.await(10, SECONDS);
+            stopwatch.stop();
+
+            assertTrue(stopwatch.elapsed(SECONDS) == (numberOfTicks * durationBetweenTicksInSeconds));
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/util/TestPeriodicTaskExecutor.java:61]: assertEquals

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java (hunk 4 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java:109]: Internally it's offloaded to an executor.

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+public class ResourceManagerClusterStatusSender
+        implements ClusterStatusSender
+{
+    private final DriftClient<ResourceManagerClient> resourceManagerClient;
+    private final InternalNodeManager internalNodeManager;
+    private final Supplier<NodeStatus> statusSupplier;
+    private final ScheduledExecutorService executor;
+    private final ResourceManagerConfig resourceManagerConfig;
+
+    private final ConcurrentMap<QueryId, PeriodicTaskExecutor> queries = new ConcurrentHashMap<>();
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java:49]: nit `Map<QueryId, PeriodicTaskExecutor> queries = new ConcurrentHashMap<>();`

**Review by highker — APPROVED** [review]

**Pull request merged**